src/Storages/S3Queue/S3QueueSource.cpp

```diff
@@ -0,0 +1,339 @@
+        {
+            if (reader)
+                reader->cancel();
+            break;
+        }
+
+        Chunk chunk;
+        bool success_in_pulling = false;
+        String file_path;
+        try
+        {
+            if (reader->pull(chunk))
+            {
+                UInt64 num_rows = chunk.getNumRows();
+
+                file_path = reader.getPath();
+                size_t total_size = file_iterator->getTotalSize();
+                if (num_rows && total_size)
+                {
+                    updateRowsProgressApprox(
+                        *this, chunk, total_size, total_rows_approx_accumulated, total_rows_count_times, total_rows_approx_max);
+                }
+
+                for (const auto & virtual_column : requested_virtual_columns)
+                {
+                    if (virtual_column.name == "_path")
+                    {
+                        chunk.addColumn(virtual_column.type->createColumnConst(num_rows, file_path)->convertToFullColumnIfConst());
+                    }
+                    else if (virtual_column.name == "_file")
+                    {
+                        size_t last_slash_pos = file_path.find_last_of('/');
+                        auto column = virtual_column.type->createColumnConst(num_rows, file_path.substr(last_slash_pos + 1));
+                        chunk.addColumn(column->convertToFullColumnIfConst());
+                    }
+                }
+                success_in_pulling = true;
+            }
+        }
+        catch (const Exception & e)
+        {
+            LOG_ERROR(log, "Exception in chunk pulling: {} ", e.displayText());
+            const auto & failed_file_path = reader.getPath();
+            queue_holder->markFailedAndCheckRetry(failed_file_path);
+            success_in_pulling = false;
+        }
+        if (success_in_pulling)
+        {
+            applyActionAfterProcessing(file_path);
+            queue_holder->setFileProcessed(file_path);
+            return chunk;
+        }
+
+
+        assert(reader_future.valid());
+        reader = reader_future.get();
+
+        if (!reader)
+            break;
+        /// Even if task is finished the thread may be not freed in pool.
+        /// So wait until it will be freed before scheduling a new task.
+        internal_source->create_reader_pool.wait();
+        reader_future = internal_source->createReaderAsync();
+    }
+
+    return {};
+}
+
+
+void StorageS3QueueSource::applyActionAfterProcessing(const String & file_path)
+{
+    switch (action)
+    {
+        case S3QueueAction::DELETE:
+            deleteProcessedObject(file_path);
+            break;
+        case S3QueueAction::KEEP:
+            break;
+    }
+}
+
+void StorageS3QueueSource::deleteProcessedObject(const String & file_path)
+{
+    LOG_WARNING(log, "Delete processed file {} from bucket {}", file_path, bucket);
+    S3::DeleteObjectRequest request;
+    String delete_key = file_path.substr(bucket.length() + 1);
```

Comment:
did not understand, why we slice bucket size bytes form the beginning? bucket can be places in any place of url, not just at the end.